k8s/src/main/scala/io/buoyant/k8s/ServiceNamer.scala

```diff
@@ -0,0 +1,174 @@
+package io.buoyant.k8s
+
+import com.twitter.conversions.time._
+import com.twitter.finagle.{Service => _, _}
+import com.twitter.finagle.service.Backoff
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.util._
+import io.buoyant.k8s.v1._
+import java.net.InetSocketAddress
+
+/**
+ * Accepts names in the form:
+ *   /<namespace>/<port-name>/<svc-name>/residual/path
+ *
+ * and attempts to bind an Addr by resolving to the external load balancer
+ * for the given service and port.
+ */
+class ServiceNamer(
+  idPrefix: Path,
+  mkApi: String => NsApi,
+  backoff: Stream[Duration] = Backoff.exponentialJittered(10.milliseconds, 10.seconds)
+)(implicit timer: Timer = DefaultTimer.twitter) extends Namer {
+
+  private[this] val PrefixLen = 3
+
+  override def lookup(path: Path): Activity[NameTree[Name]] = path.take(PrefixLen) match {
+    case id@Path.Utf8(nsName, portName, serviceName) =>
+
+      val nameTree = serviceNs.get(nsName).get(serviceName, portName).map {
+        case Some(address) =>
+          val bound = address.map(Addr.Bound(_))
+          NameTree.Leaf(Name.Bound(bound, idPrefix ++ id, path.drop(PrefixLen)))
+        case None =>
+          NameTree.Neg
+      }
+      Activity(nameTree.map(Activity.Ok(_)))
+
+    case _ =>
+      Activity.value(NameTree.Neg)
+  }
+
+  private[this] def getPort(service: Service, portName: String): Option[Int] =
+    for {
+      spec <- service.spec
+      port <- spec.ports.find(_.name == portName)
+    } yield port.port
+
+  val serviceNs = new Ns[Service, ServiceWatch, ServiceList, ServiceCache](backoff, timer) {
```

Comment:
does this need to be public?

also, prefer type annotation like:
```scala
val serviceNs: Ns[Service, ServiceWatch, ServiceList, ServiceCache] = new Ns(backoff, timer {
```

I care less about this if it's not public.